### ISSUE TITLE
DaemonSet dynamic/scaled resources based on node type

### ISSUE THREAD

**Issue by wiardvanrij** [kubernetes/kubernetes#122753]
### What would you like to be added?

_Intro note: I've read about KEP, however, I'm merely a busy engineer that loves Kubernetes. I'm not sure I can lead my feature request through the KEP process. Please consider this as my initial reach out, if anything to have my voice heard and poll my idea to the community. If there is enough traction, I could follow up and/or someone else could lead it._

I want to propose an enhancement to the DaemonSet (DS) resource management in Kubernetes. The DaemonSet, as it stands, is an invaluable feature for ensuring that each node in a cluster runs a copy of a specific pod. However, I have identified an area where its functionality could be significantly improved: dynamic and scaled resource allocation based on node type.

### Current Limitation: Static Resource Definition

Currently, the resource allocation logic for DaemonSets is rather static. When defining resources for a DaemonSet, such as CPU and memory, it operates on an 'all-or-nothing' basis. This approach does not leverage the full potential of diverse node environments.

### Diverse Node Environment in Current Ecosystem

In the present cloud ecosystem, there's an extensive variety of node types available. These range from nodes with as little as 1 virtual CPU to those as large as 488 CPUs offered by services like AWS. This diversity in node capabilities is not currently accounted for in the DaemonSet design.

### Resource Dependence on Node Size and Pod Density

The workload resources for a DaemonSet, particularly memory and CPU, are often dependent on the number of pods running on a node. More importantly, they can also depend on the size or capacity of the node itself. For instance, a node with more CPUs and memory can, and ideally should, handle more intensive workloads.

### Lack of Native Methods for Dynamic Resource Allocation

As of now, Kubernetes does not offer native methods to define resources for a DaemonSet based on the size or capacity of the node. While it is possible to categorize nodes into groups based on size, this solution is cumbersome and static. It requires significant manual intervention and does not adapt dynamically as the cluster evolves.

### Existing Alternatives

I acknowledge that there are existing methods to manage this challenge, such as adding nodes to different node groups and then using affinity rules to place DaemonSets with different resource requirements. However, this approach is quite tedious and complex. It requires significant manual configuration and continuous adjustment, which detracts from usability and efficiency. This method does not truly cater to the dynamic nature of resource demands in a Kubernetes environment.

### Proposal: Dynamic/Scaled Resource Allocation for DaemonSets

I propose the introduction of a feature that allows DaemonSets to dynamically allocate resources based on the type or capacity of the node. This feature would analyze the node's characteristics (such as CPU andmemory capacity plus it's maximum pods that it can hold) and automatically define the resources allocated to the DaemonSet's pods running on that node.

This dynamic allocation would ensure more efficient use of cluster resources, improve performance, and potentially reduce costs by maximizing the utilization of each node's capacity. It would also significantly reduce the management overhead for Kubernetes administrators, as the system would automatically adapt to changes in the node landscape.

In summary, incorporating dynamic and scaled resource allocation for DaemonSets based on node type would be a substantial improvement in Kubernetes' functionality. It would enhance the adaptability of DaemonSets to diverse and evolving node environments, leading to more efficient resource usage and simplified cluster management.

### Why is this needed?

^

**Comment by thockin** [comment]
This topic is not new, but it is hard.  We've discussed a `DaemonSetSet` type which could use labels to chose between N DaemonSet configs.   We've discussed making percentage resources to claim x% of a node.  The design for this is very unclear, but I wanted to add some color.

I'm not going to go hunting now, but I'd be shocked if this is not a dup :)

**Comment by wiardvanrij** [comment]
> This topic is not new, but it is hard. We've discussed a `DaemonSetSet` type which could use labels to chose between N DaemonSet configs. We've discussed making percentage resources to claim x% of a node. The design for this is very unclear, but I wanted to add some color.
> 
> I'm not going to go hunting now, but I'd be shocked if this is not a dup :)

I've done some searching and I recall seeing an answer from you on a forum as well. I haven't really seen a more 'formal' request for it though.

Would you recommend I start something with KEP? I could potentially create some priority (meaning time) within our company to guide this into a real suggestion / improvement.

**Comment by thockin** [comment]
Before jumping into a full KEP, I would do a "pre-KEP" as a simple GoogleDoc or something to collect options and get feedback on tradeoffs, both for implementation options (e.g. built-in vs. CRD) and for the overall model (e.g. DaemonSetSet or percentage or ...)

I suspect thi sis sig-apps owned, so maybe soliciti some early feedback there.

I find this to be the most iterative approach to such a significant proposal :)

**Comment by garvinp-stripe** [comment]
Wanted to throw in random ideas since we struggled with similar problems not just with daemonsets but also sidecars. I think fundamentally we are trying to solve mapping resource requirements of containers to some kind of selector that is user configurable. 

VPA does this to some degree in that it understands per container utilization at pod level. However it forces target to be on the whole daemonset/ deployment/etc which isn't what we want. In addition, we want VPA's ability to take over resource requirement on pod launch. 

Following VPA's foot step. If we can introduce a new CRD that similar to VPA, or maybe take over the delegation of resource requirement from VPA to this CRD, that allow users to specify targeting via a selector and resource requirement that would likely be ideal?

**Comment by bradmorgtile** [comment]
Running into this issue when trying to run clusters with a diverse variety of nodetypes. Very difficult to make those smaller nodes efficient when you have to statically assign requests.

**Comment by trois-six** [comment]
Hello there, you can find a solution to this problem [here](https://github.com/ManoManoTech/kubernetes-node-specific-sizing). IMO, this should be directly included in the Daemon set spec.

**Comment by thockin** [comment]
I still think we need something here, but the exact shape of it isn't clear.  As discussed up-thread, there are several ways to approach it, and even more options as we evolve.

We could maybe use CEL - provide an expression based on the node's resources for each resource you care about, use the result of that expression.

We could maybe rely on in-place pod updates. 

So I will reopen this, but at the moment it does not have an active owner. :(

**Comment by trois-six** [comment]
We made a first approach with static pourcentages of the node resources, but we want to work on a CEL based implementation in the future.

**Comment by eliebleton-manomano** [comment]
Hello! I'm one of the authors of the solution trois-six mentioned.

> We could maybe rely on in-place pod updates.

We actually started there, using an actual CR to specify resource allocation instead of a pod mutating hook. This being in alpha makes it tougher to roll-out at this time, which motivated us to write the mutating hook. 

The hook can only work for pods that are pinned to a node somehow before scheduling - that's only DaemonSet-controlled pods as far as I'm aware. In contrast, the CR+in-place updates could work with any Pod. I struggle to find a legitimate use-case for node-sized pods that would be controlled by a ReplicaSet or a Job.

On the topic on in-place pod updates, it seems designed to dramatically improve VPA, and after writing the first node-specific-sizing controller using in-place updates, my thinking was that VPA would probably do the job just as well, and that an in-place node-specific-sizing is not very enticing.

> We could maybe use CEL - provide an expression based on the node's resources for each resource you care about, use the result of that expression.

Exactly my thoughts! It would open up the door for changing behavior depending on labels, or NodeGroup, or anything else. Or a different splitting algorithm. I think there is a tradeoff between configuration intelligibility and expressiveness in the CEL implementation.

We opted to go for a comparatively simpler proportional sizing which I believe that will help us measure the expected benefits in our situation. I'm not convinced we will need CEL to achieve our goals, but I'm looking forward implementing some of it as an experiment.

---

On that experimental note, we published the code not with the hopes it would solve everybody's problem even in time, but rather with the hopes it would restart this very conversation thread, allow people to see for themselves if node-specific-sizing is a solution or not to the problems they have, and also maybe be a test-bed for different ways to go about this.

> this should be directly included in the DaemonSet spec.

I don't know that our approach _should_ be included, but I know that it _could_ : it really just mutates pod specs that are created by the DaemonSet controller. As such, if the DaemonSet was creating them differently, it would do produce the same effect at less cost for the cluster.

**Comment by thockin** [comment]
> I struggle to find a legitimate use-case for node-sized pods that would be controlled by a ReplicaSet or a Job.

A lot of AI workloads are "give me the whole machine".  

The benefit of describing this as a daemonset idea is that DS knows the node BEFORE scheduling.  Doing proportional resources AFTER scheduling (or during) is much more complicated.

**Comment by pznamensky** [comment]
@thockin can you please reopen the issue?

**Comment by mlladb** [comment]
I know this issue is closed but I'm trying to figure out the best approach for dynamic resources requests based on node type for DaemonSets. Is there another open issues I can track?

What's the recommended approach?
- The mutating admission controller mentioned in the comments [here](https://github.com/kubernetes/kubernetes/issues/122753#issuecomment-2351142332).
- A variation on VPA mentioned in the comments [here](https://github.com/kubernetes/kubernetes/issues/122753#issuecomment-1984814453)
- A separate dedicated DaemonSet for each NodePool mentioned in the original post.
- Something else?

**Comment by afreyermuth98** [comment]
Now that the 1.33 has been released. Do you have an idea / example on how to do this ?

**Comment by mlladb** [comment]
> Now that the 1.33 has been released. Do you have an idea / example on how to do this ?

If this relates to this comment https://github.com/kubernetes/kubernetes/issues/122753#issuecomment-2764879412 the [preview docs](https://deploy-preview-50290--kubernetes-io-main-staging.netlify.app/docs/tasks/configure-pod-container/resize-container-resources/) suggest the ability to patch resource requests as follows. 

```bash
kubectl patch pod resize-demo --subresource resize --patch \
  '{"spec":{"containers":[{"name":"pause", "resources":{"requests":{"cpu":"800m"}, "limits":{"cpu":"800m"}}}]}}'
```

In a production environment its probably not plausible to patch individual pods. An operator could monitor pod resource utilisation and trigger resizing dynamically but to the best of my knowledge that does not exist and would have to be created.

**Comment by mlladb** [comment]
Is there an alternative issue to track this feature request?
